src/cmap/cmap.go

```diff
@@ -130,6 +130,25 @@ func (s *shard[K, V]) Set(key K, val V, overwrite bool) (V, bool) {
 	return old, true
 }
 
+// LazySet is like Set but calls the given function to construct the object only if needed.
+func (s *shard[K, V]) LazySet(key K, f func() V) (V, bool) {
+	s.l.Lock()
+	defer s.l.Unlock()
+	if existing, present := s.m[key]; present {
+		if existing.Wait == nil {
+			return existing.Val, false // already added
+		}
+		// Hasn't been added, but something is waiting for it to be.
+		s.m[key] = awaitableValue[V]{Val: f()}
+		close(existing.Wait)
+		existing.Wait = nil
+		return existing.Val, true
```

Comment:
Should this be `s.m[key].Val`? IIUC, `existing.Val` will be zero if `existing.Wait` is non-nil.

It looks like the existing tests don't actually check the return value of `Set` (and I question whether it actually needs to return the value at all)

Reply:
`Set` was supposed to return the old value; there'd be no point in returning `s.m[key].Val` there since that's what the caller supplied.

I think you're right that nothing actually cares about this any more though, so maybe it can all just be simplified out.

Reply:
Okay right, I've removed it from Set where it was no longer needed. This now returns whatever is set now which is more useful (the callers of it never look at the value in that code path but it still makes more sense this way).